CHANGES.md

```diff
@@ -24,6 +24,11 @@ Tactics
   Simplex-based proof engine. In case of regression, 'Unset Simplex'
   to get the venerable Fourier-based engine.
 
+Diffs
+
+- Some error messages that show problems with a pair of non-matching values will now
+  highlight the differences (when diffs are enabled).
```

Comment:
I am not sure what is the point of requiring the user to turn diffs on (most users probably won't most of the time) to see diffs in error messages. It seems useful even to users who do not want diffs between successive goals (and it definitely sounds like two unrelated things).

Reply:
A problem with the diff code is that it likes to die on quite a few inputs, it seems.

Reply:
Yes but if there are such issues we will be more likely to detect them before 8.10 if this is activated by default. If this code path is never used in our tests, it is more likely to bitrot quickly.

Reply:
IMHO the code needs significant work before is activated by default; also, it has the potential to increase compile time largely, so I guess we would be best served by specific tests.

Reply:
In any case, it doesn't make sense to use the same option for diffs between successive goals and diffs in error messages, because the former should never be made the default while the latter should (as soon as it is realistic to do so).

Reply:
> I am not sure what is the point of requiring the user to turn diffs on (most users probably won't most of the time) to see diffs in error messages.

That makes sense; I would do that, though I would hope that users turn on diffs more often.  In coqtop, we would need to check that `-color on` was specified on the command line.

> A problem with the diff code is that it likes to die on quite a few inputs, it seems.

No.  I'm pretty careful about not breaking things for users:
1) The values that are being compared in proof steps and in messages are terms and are unlikely to include the few input sequences that are not handled yet.  https://github.com/coq/coq/pull/8734 addresses those.
2) In the unlikely case that the diff fails, it prints the message without diff highlighting and an additional message saying that diff failed.  The user impact is negligible.

> I guess we would be best served by specific tests.

Automatic testing is an interesting challenge.  I included a .v file that generates an example message (for a manual test).  Do we have a mechanism in place to verify that coqtop output contains the message?

> IMHO the code needs significant work before is activated by default

Specifics please?

> it has the potential to increase compile time largely

Huh?